pkg/agent/openflow/packetin.go

```diff
@@ -15,63 +15,101 @@
 	handler, ok := packetInHandler.(PacketInHandler)
 	if !ok {
-		klog.Errorf("Invalid Traceflow controller.")
+		klog.Errorf("Invalid controller to handle packetin.")
 		return
 	}
-	c.packetInHandlers[packetHandlerName] = handler
+	if c.packetInHandlers[packetHandlerReason] == nil {
+		c.packetInHandlers[packetHandlerReason] = map[string]PacketInHandler{}
+	}
+	c.packetInHandlers[packetHandlerReason][packetHandlerName] = handler
 }
 
-func (c *client) StartPacketInHandler(stopCh <-chan struct{}) {
-	if len(c.packetInHandlers) == 0 {
-		return
-	}
-	ch := make(chan *ofctrl.PacketIn)
-	err := c.SubscribePacketIn(uint8(ofprAction), ch)
-	if err != nil {
-		klog.Errorf("Subscribe PacketIn failed %+v", err)
-		return
-	}
-	packetInQueue := workqueue.NewNamed("packetIn")
-	go c.parsePacketIn(packetInQueue, stopCh)
+type FeatureStartPacketIn interface {
+	ConfigPacketIn()
+	ListenPacketIn()
+}
+
+type featureStartPacketIn struct {
+	reason        uint8
+	subscribeCh   chan *ofctrl.PacketIn
+	stopCh        <-chan struct{}
+	packetInQueue *workqueue.Type
+}
+
+func (f *featureStartPacketIn) ConfigPacketIn() {
+	f.subscribeCh = make(chan *ofctrl.PacketIn)
+	f.packetInQueue = workqueue.NewNamed(string(f.reason))
+}
 
+func (f *featureStartPacketIn) ListenPacketIn() {
 	for {
 		select {
-		case pktIn := <-ch:
+		case pktIn := <-f.subscribeCh:
 			// Ensure that the queue doesn't grow too big. This is NOT to provide an exact guarantee.
-			if packetInQueue.Len() < packetInQueueSize {
-				packetInQueue.Add(pktIn)
+			if f.packetInQueue.Len() < packetInQueueSize {
+				f.packetInQueue.Add(pktIn)
 			} else {
 				klog.Warningf("Max packetInQueue size exceeded.")
 			}
-		case <-stopCh:
-			packetInQueue.ShutDown()
-			break
+		case <-f.stopCh:
+			f.packetInQueue.ShutDown()
+			return
+		}
+	}
+}
+
+func (c *client) StartPacketInHandler(packetInStartedReason []uint8, stopCh <-chan struct{}) {
+	if len(c.packetInHandlers) == 0 || len(packetInStartedReason) == 0 {
+		return
+	}
+
+	// Iterate through each feature that starts packetin. Subscribe with their specified reason.
+	for _, reason := range packetInStartedReason {
+		featurePacketIn := &featureStartPacketIn{reason: reason, stopCh: stopCh}
+		err := c.subscribeFeaturePacketIn(featurePacketIn)
+		if err != nil {
+			klog.Errorf("received error %+v while subscribing packetin for each feature", err)
 		}
 	}
 }
 
-func (c *client) parsePacketIn(packetInQueue workqueue.Interface, stopCh <-chan struct{}) {
+func (c *client) subscribeFeaturePacketIn(featurePacketIn *featureStartPacketIn) error {
+	featurePacketIn.ConfigPacketIn()
+	err := c.SubscribePacketIn(featurePacketIn.reason, featurePacketIn.subscribeCh)
+	if err != nil {
+		return errors.New(fmt.Sprintf("Subscribe %d PacketIn failed %+v", featurePacketIn.reason, err))
+	}
+	go c.parsePacketIn(featurePacketIn.packetInQueue, featurePacketIn.reason)
+	featurePacketIn.ListenPacketIn()
```

Comment:
`ListenPacketIn` is a blocking method that never returns unless `stopCh` is closed while `subscribeFeaturePacketIn` is called in a loop, so the `subscribeFeaturePacketIn` for the second reason is not going to be called.

Reply:
Thanks, putting it in a goroutine.